hpx/lcos/broadcast_async.hpp

```diff
@@ -0,0 +1,211 @@
+//  Copyright (c) 2013-2017 Hartmut Kaiser
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+/// \file broadcast_async.hpp
+
+#ifndef HPX_LCOS_BROADCAST_ASYNC_HPP
+#define HPX_LCOS_BROADCAST_ASYNC_HPP
+
+#include <hpx/config.hpp>
+#include <hpx/async.hpp>
+#include <hpx/lcos/dataflow.hpp>
+#include <hpx/lcos/future.hpp>
+#include <hpx/lcos/promise.hpp>
+#include <hpx/lcos/when_all.hpp>
+#include <hpx/runtime/actions/plain_action.hpp>
+#include <hpx/runtime/agas/symbol_namespace.hpp>
+#include <hpx/runtime/basename_registration.hpp>
+#include <hpx/runtime/naming/id_type.hpp>
+#include <hpx/runtime/trigger_lco.hpp>
+#include <hpx/traits/acquire_shared_state.hpp>
+#include <hpx/traits/future_access.hpp>
+#include <hpx/util/bind.hpp>
+#include <hpx/util/deferred_call.hpp>
+#include <hpx/util/calculate_fanout.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <map>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include <boost/format.hpp>
+#include <boost/preprocessor/cat.hpp>
+
+#if !defined(HPX_BROADCAST_FANOUT)
+#define HPX_BROADCAST_FANOUT 16
+#endif
+
+namespace hpx { namespace lcos
+{
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename T>
+    hpx::future<T> broadcast_recv(char const* basename,
+        std::size_t this_site = std::size_t(-1),
+        std::size_t generation = std::size_t(-1))
+    {
+        if (this_site == std::size_t(-1))
+            this_site = static_cast<std::size_t>(hpx::get_locality_id());
+
+        std::string name(basename);
+        if (generation != std::size_t(-1))
+            name += std::to_string(generation) + "/";
+
+        typedef typename std::decay<T>::type result_type;
+
+        // this is the receiving endpoint for this site
+        hpx::lcos::promise<result_type> p;
+        hpx::future<result_type> f = p.get_future();
+
+        // register promise using symbolic name
+        hpx::future<bool> was_registered =
+            hpx::register_with_basename(name, p.get_id(), this_site);
+
+        return hpx::dataflow(
+            [](hpx::future<result_type> f, hpx::future<bool> was_registered,
+                std::string && name, std::size_t this_site) -> result_type
+            {
+                // rethrow exceptions
+                was_registered.get();
+
+                // make sure promise gets unregistered after use
+                hpx::unregister_with_basename(name, this_site).get();
+
+                // propagate result
+                return f.get();
```

Comment:
Why? We don't care about the result of the `unregister_with_basename` and nothing depends on it.